File: gui_apps/ch01/sigslot2/sigslot2.py
# sigslot.py: a more realistic signals/slots demo
import sys
import pathlib
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

USING_PYQT6 = True if PYQT_VERSION_STR.startswith('6') else False

p = pathlib.Path(__file__)
print(f"Current file's path is {str(p)}")
p.parents[2]
print(f"p.parents[2] = {str(p.parents[2])}")
sys.path.append(str(p.parents[2]))
import globalvars
globalvars.USING_PYQT6 = USING_PYQT6
import utils


class SigSlotWidget(utils.CenteredOnDesktopWidget):
    def __init__(self, *args, **kwargs):
        super(SigSlotWidget, self).__init__(*args, **kwargs)

        # setup Gui
        self.setWindowTitle("PyQt: Signals & Slots - 2")

        # build the layout
        # top row - just the prompt
        self.prompt_label = QLabel("Spin the spinbox below & see the " +
                                   "slider and label getting updated")
        # middle row: spinner - slider - label
        self.spinner = QSpinBox()
        self.spinner.setRange(-50, 50)
        self.slider = QSlider(Qt.Horizontal)
        self.slider.setRange(-50, 50)
        self.val_label = QLabel("")
        # bottom row - hgap - push button
        self.btnQuit = QPushButton("Quit")

        h_layout1 = QHBoxLayout()
        h_layout1.addWidget(self.spinner)
        h_layout1.addWidget(self.slider)
        h_layout1.addWidget(self.val_label)

        h_layout2 = QHBoxLayout()
        h_layout2.addStretch()
        h_layout2.addWidget(self.btnQuit)

        v_layout = QVBoxLayout()
        v_layout.addWidget(self.prompt_label)
        v_layout.addLayout(h_layout1)
        v_layout.addLayout(h_layout2)
        self.setLayout(v_layout)

        # setup signals & slots
        self.spinner.valueChanged.connect(self.val_label.setNum)
        self.spinner.valueChanged.connect(self.slider.setValue)
        self.btnQuit.clicked.connect(QApplication.instance().quit)

        # kick-off all the signals & slots
        self.spinner.setValue(5)
        self.spinner.setValue(0)


def main():
    app = QApplication(sys.argv)
    # font = QFont(app.font().family(), 12)
    # app.setFont(font)
    app.setFont(QApplication.font("QMenu"))

    w = SigSlotWidget()
    w.show()
    return app.exec()


if __name__ == '__main__':
    main()

File: gui_apps/ch01/sigslot2/sigslot2.cc
// ch01/sigslot2/sigslot2.cc: illustrates a more realistic signal/slot example
#include <QApplication>
#include <QDesktopWidget>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QSlider>
#include <QSpinBox>
#include <QVBoxLayout>

void centerWidgetOnDesktop(QWidget &window)
{
   window.show();
   QRect screenSize = QApplication::desktop()->screenGeometry();
   int x = (screenSize.width() - window.width()) / 2;
   int y = (screenSize.height() - window.height()) / 2;
   window.move(x, y);
}

int main(int argc, char **argv)
{
   // @see: https://doc.qt.io/archives/qt-5.6/qtlabscontrols-highdpi.html
   QGuiApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
   QApplication app(argc, argv);
   QFont font("Segoe UI", 12);
   QApplication::setFont(font);

   // build your GUI here....
   QWidget window;
   window.setWindowTitle("Qt Signal Slot Example");
   // window.setGeometry(QRect(QPoint(0, 0), QSize(450, 80)));

   // setup UI elements
   // top-row: label prompt
   QLabel *prompt = new QLabel("Spin the spinbox below & see the "
                               "slider and label getting updated.");
   // middle row: spinner - slider - label
   QSpinBox *spinBox = new QSpinBox();
   spinBox->setRange(-50, 50);
   QSlider *slider = new QSlider(Qt::Horizontal);
   slider->setRange(-50, 50);
   QLabel *label = new QLabel("0");
   // bottom row: stretch + pushbutton
   QPushButton *btnQuit = new QPushButton("Quit");

   // middle row
   QHBoxLayout *hl = new QHBoxLayout();
   hl->addWidget(spinBox);
   hl->addWidget(slider);
   hl->addWidget(label);
   // bottom row
   QHBoxLayout *h2 = new QHBoxLayout();
   h2->addStretch();
   h2->addWidget(btnQuit);
   // main layout
   QVBoxLayout *layout = new QVBoxLayout();
   layout->addWidget(prompt);
   layout->addLayout(hl);
   layout->addLayout(h2);
   window.setLayout(layout);

   // setup signals & slots
   QObject::connect(spinBox, SIGNAL(valueChanged(int)), label, SLOT(setNum(int)));
   QObject::connect(spinBox, SIGNAL(valueChanged(int)), slider, SLOT(setValue(int)));
   QObject::connect(slider, SIGNAL(valueChanged(int)), label, SLOT(setNum(int)));
   QObject::connect(slider, SIGNAL(valueChanged(int)), spinBox, SLOT(setValue(int)));
   QObject::connect(btnQuit, SIGNAL(clicked()), qApp, SLOT(quit()));

   // this will kick off the signals
   spinBox->setValue(0);
   centerWidgetOnDesktop(window);

   return app.exec();
}
